xgboost_ray/main.py

```diff
@@ -342,27 +345,35 @@ def _train(params: Dict,
     # Train
     fut = [
         actor.train.remote(rabit_args, params, dtrain, evals, *args, **kwargs)
-        for (actor, _) in actors
+        for actor in actors
     ]
 
     callback_returns = [list() for _ in range(len(actors))]
     try:
-        ready, not_ready = ray.wait(fut, num_returns=len(fut), timeout=0)
+        not_ready = fut
         while not_ready:
-            for i, (actor, queue) in enumerate(actors):
+            if queue:
                 while not queue.empty():
                     item = queue.get()
                     if isinstance(item, Callable):
                         item()
                     else:
                         callback_returns[i].append(item)
-            ready, not_ready = ray.wait(fut, num_returns=len(fut), timeout=0)
+            ready, not_ready = ray.wait(not_ready, timeout=0)
+            logger.info("[RayXGBoost] Spinning...")
+            time.sleep(1)
+            ray.get(ready)
         # Once everything is ready
         ray.get(fut)
     except RayActorError:
-        for (actor, _) in actors:
+        for actor in actors:
             ray.kill(actor)
+        if queue:
+            ray.kill(queue)
```

Comment:
this is also important to kill, otherwise we end up leaking a bunch of queues (which may not necessarily garbage collect properly).

Reply:
Good spot thanks.
Should we catch broader here?

Reply:
I'll update the code below